src/mesh/p4est_mesh.jl

```diff
@@ -0,0 +1,273 @@
+
+An unstructured curved mesh based on trees, which uses the C library p4est
+to manage trees and mesh refinement.
+
+!!! warning "Experimental code"
+    This mesh type is experimental and can change any time.
+"""
+mutable struct P4estMesh{NDIMS, RealT<:Real, NDIMSP2} <: AbstractMesh{NDIMS}
+  p4est                 ::Ptr{p4est_t}
+  p4est_mesh            ::Ptr{p4est_mesh_t}
+  tree_node_coordinates ::Array{RealT, NDIMSP2} # [dimension, i, j, k, tree_id]
+  nodes                 ::Vector{RealT}
+  periodicity           ::NTuple{NDIMS, Bool}
+  current_filename      ::String
+  unsaved_changes       ::Bool
+end
+
+
+"""
+    P4estMesh(trees_per_dimension, mapping, nodes::AbstractVector;
+              RealT=Float64, initial_refinement_level=0, periodicity=true, unsaved_changes=true)
+
+Create a P4estMesh that is a structured curved hypercube of the specified size.
+
+# Arguments
+- `trees_per_dimension::NTupleE{NDIMS, Int}`: the number of trees in each dimension.
+- `mapping`: a function of `NDIMS` variables to describe the mapping, which transforms
+             the reference mesh (`[-1, 1]^n`) to the physical domain.
+- `nodes::AbstractVector`: nodes which should be used to store the geometry of the mesh.
+                           The mapping will be evaluated at the tensor product of the specified nodes
+                           for each tree and will then be interpolated for refined cells.
+- `RealT::Type`: the type that should be used for coordinates.
+- `initial_refinement_level::Integer`: refine the mesh uniformly to this level before the simulation starts.
+- `periodicity`: either a `Bool` deciding if all of the boundaries are periodic or an `NTuple{NDIMS, Bool}`
+                 deciding for each dimension if the boundaries in this dimension are periodic.
+- `unsaved_changes::Bool`: if set to `true`, the mesh will be saved to a mesh file.
+"""
+function P4estMesh(trees_per_dimension, mapping, nodes::AbstractVector;
+                   RealT=Float64, initial_refinement_level=0, periodicity=true, unsaved_changes=true)
+  NDIMS = length(trees_per_dimension)
+
+  # Convert periodicity to a Tuple of a Bool for every dimension
+  if all(periodicity)
+    # Also catches case where periodicity = true
+    periodicity = ntuple(_->true, NDIMS)
+  elseif !any(periodicity)
+    # Also catches case where periodicity = false
+    periodicity = ntuple(_->false, NDIMS)
+  else
+    # Default case if periodicity is an iterable
+    periodicity = Tuple(periodicity)
+  end
+
+  tree_node_coordinates = Array{RealT, NDIMS+2}(undef, NDIMS,
+                                                ntuple(_ -> length(nodes), NDIMS)...,
+                                                prod(trees_per_dimension))
+  calc_node_coordinates!(tree_node_coordinates, mapping, trees_per_dimension, nodes)
+
+  # p4est_connectivity_new_brick has trees in Morton order, so use our own function for this
+  conn = connectivity_structured_periodic(trees_per_dimension...)
```

Comment:
The function `calc_node_coordinates!` in `p4est_mesh.jl`, which computes the coordinates of the trees, would be a lot more complicated (I guess).

Reply:
OK, but this affects only these two functions, right? Thus if we wanted to change the sorting later, we could just do it, since this sorting is not implicitly used anywhere else?

Reply:
Yes, in general, the numbering of trees is irrelevant, but we need to compute the node coordinates somehow in Trixi, which is easier with this numbering.

Reply:
OK, then leave it as it is, but maybe document the fact that this can be changed independently in the code for future reference